### PR TITLE
Closing loop

### PR DESCRIPTION
**Pull request by karim-alweheshy**

### CODE REVIEW

**Comment by NadaChatti** [comment]
I thought we agreed to make PRs for each branch and NOT use this one? This has a LOT of stuff that should not be in the develop branch..

**Comment by mlanas** [comment]
> I thought we agreed to make PRs for each branch and NOT use this one? This has a LOT of stuff that should not be in the develop branch..

Last meeting we agreed to merge this one, so the idea is to clean this branch for merging with ss20_devel

File: bulletroboy/exoforce_simulation.py [hunk 1]
```diff
@@ -23,19 +23,21 @@ def __init__(self, cage_conf, human_model, mode):
...
 		if self.mode == "debug":
 			self.init_debug_parameters()
 		else:
 			if self.mode == "tendon":
 				self.create_subscription(TendonUpdate, '/roboy/simulation/tendon_force', self.tendon_update_listener, 10)
 			elif self.mode == "forces":
-				# TODO: update subscriber with correct msg type
-				self.create_subscription(TendonUpdate, '/roboy/simulation/operator_forces', self.forces_update_listener, 10)
+				self.get_logger().info('forces')
+				self.create_subscription(Collision, '/roboy/exoforce/collisions', self.collision_listener, 10)
```
Comment (missxa) [inline bulletroboy/exoforce_simulation.py]: decrease queue size to 1

File: bulletroboy/exoforce_simulation.py [hunk 2]
```diff
@@ -23,19 +23,21 @@ def __init__(self, cage_conf, human_model, mode):
...
 			self.create_subscription(Float32, '/roboy/simulation/cage_rotation', self.cage_rotation_listener, 10)
```
Comment (missxa) [inline bulletroboy/exoforce_simulation.py]: decrease queue size to 1

File: bulletroboy/exoforce_simulation.py [hunk 3]
```diff
@@ -64,18 +66,39 @@ def init_debug_parameters(self):
...
 
 	def tendon_update_listener(self, tendon_force):
 		self.update_tendon(tendon_force.tendon_id, tendon_force.force)
 
 	def cage_rotation_listener(self, angle):
 		self.rotate_cage(angle.data)
+	
+
+	def collision_listener(self, collision):
+		self.get_logger().info('Received collision message in roboy link %i' % collision.linkid)
```
Comment (missxa) [inline bulletroboy/exoforce_simulation.py]: throttle this log, limit to once per second or per 5 seconds

File: bulletroboy/exoforce_simulation.py (hunk 1 above)
Comment (missxa) [inline bulletroboy/exoforce_simulation.py]: `/roboy/simulation/(exoforce)/operator/collisions`is a more suitable name

File: bulletroboy/exoforce_simulation.py [hunk 4]
```diff
@@ -64,18 +66,39 @@ def init_debug_parameters(self):
...
 
-	def forces_update_listener(self, forces):
-        # TODO: implement force update
-		pass
+		if visualize == True:
+			# While the force is being applied in link frame with p.applyexternalforce, the debuglines can only
+			# be made in world frame, hence visualization needs link to frame orientation and position mapping.
+			# TODO: Once link orientation map is implemented, use it to pass force_end to world coordinates!
+			force_end = [(position_on_link[i] - force_vec[i]) for i in range(len(position_on_link))]
+			print('force_end: ', force_end)
```
Comment (missxa) [inline bulletroboy/exoforce_simulation.py]: no prints please

File: bulletroboy/forces_mapper.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        super().__init__('forces_imitator')
+
+        self.action_done_event = Event()
+
+        self.callback_group = ReentrantCallbackGroup()
+        self.link_names_map = self.load_roboy_to_human_link_name_map()
+
+        self.robotCollisionSubscription = self.create_subscription(
+            Collision,
+            '/roboy/simulation/collision',
```
Comment (missxa) [inline bulletroboy/forces_mapper.py]: `roboy/simulation/roboy/collision`?

File: bulletroboy/forces_mapper.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            self.collision_listener,
+            10)
```
Comment (missxa) [inline bulletroboy/forces_mapper.py]: decrease queue size to 1

File: bulletroboy/forces_mapper.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+        # Define clients
+        self.roboy_link_info_from_id_client = self.create_client(LinkInfoFromId, '/roboy/simulation/roboy/link_info_from_id', callback_group=self.callback_group)
+        self.operator_link_info_from_name_client = self.create_client(LinkInfoFromName, '/roboy/simulation/operator/link_info_from_name', callback_group=self.callback_group)
+        
+        # Define publishers
+        self.exoforceCollisionPublisher = self.create_publisher(Collision, '/roboy/exoforce/collisions', 10)
```
Comment (missxa) [inline bulletroboy/forces_mapper.py]: see my comment on the topic name in exoforce file

Comment (missxa) [reply bulletroboy/forces_mapper.py]: decrease queue size to 1

File: bulletroboy/forces_mapper.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    def call_service(self, client, msg):
+        while not client.wait_for_service(timeout_sec=1.0):
+            self.get_logger().info("service not available, waiting again...")
+        response = client.call(msg)
+        return response
+
+    def collision_listener(self, msg):
+        """Collision subscriber handler.
+        """
+        self.get_logger().info('Received collision')
```
Comment (missxa) [inline bulletroboy/forces_mapper.py]: throttle info prints once every 2 or 3 seconds is enough

File: bulletroboy/forces_mapper.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        self.transform_from_roboy_to_operator(msg)
+
+    def transform_from_roboy_to_operator(self, collision):
+        """Calculates the collision on the human and assign it to the 'result' var.
```
Comment (missxa) [inline bulletroboy/forces_mapper.py]: incorrect docstring

File: bulletroboy/forces_mapper.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        while not client.wait_for_service(timeout_sec=1.0):
+            self.get_logger().info("service not available, waiting again...")
+        response = client.call(msg)
+        return response
+
+    def collision_listener(self, msg):
+        """Collision subscriber handler.
+        """
+        self.get_logger().info('Received collision')
+        self.transform_from_roboy_to_operator(msg)
```
Comment (missxa) [inline bulletroboy/forces_mapper.py]: it's confusing that the function is called transform but it's publishing too.. i suggest to publish from this method

File: bulletroboy/forces_mapper.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+    def transform_from_roboy_to_operator(self, collision):
+        """Calculates the collision on the human and assign it to the 'result' var.
+
+        Parameters:
+            collision (Collision):The collision that happened on the robot side.
+        """
+        operator_collision = self.map_collision_to_operator(collision)
+        self.get_logger().info("publishing")
```
Comment (missxa) [inline bulletroboy/forces_mapper.py:103]: no need, if you want it  use `.debug()` and set an appropriate logging level

File: bulletroboy/forces_mapper.py [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+    def map_collision_to_operator(self, roboy_collision):
+        """Maps roboy link id in collision to the operator corresponding link id
+
+        Parameters:
+            roboy_collision (Collision): The roboy collision
+
+        Returns:
+            Collision: The collision with mapped linkid
+        """
+        self.get_logger().info('mapping start')
```
Comment (missxa) [inline bulletroboy/forces_mapper.py]: same, use debug if you want to keep this

Comment (missxa) [reply bulletroboy/forces_mapper.py]: wont comment on the future unnecessary logs

File: bulletroboy/forces_mapper.py [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+        collision.position.x = collision.position.x / scale[0]
+        collision.position.y = collision.position.y / scale[1]
+        collision.position.z = collision.position.z / scale[2]
+
+        return collision
+
+    def load_roboy_to_human_link_name_map(self):
+        """Fetches the link name map"""
+        script_dir = os.path.dirname(__file__)
+        rel_path = "../resource/roboy_to_human_linkname_map.yaml"
```
Comment (missxa) [inline bulletroboy/forces_mapper.py]: can you guarantee that this relative path works independent from the way i execute your code?

File: bulletroboy/operator.py [hunk 14]
```diff
@@ -3,45 +3,131 @@
...
 		self.movements = Movements(self)
+
+		self.prevPose = [0, 0, 0]
+		self.trailDuration = 5
+
 		self.ef_publisher = self.create_publisher(PoseStamped, '/roboy/simulation/operator/pose/endeffector', 10)
+		self.link_info_service = self.create_service(LinkInfoFromName, '/roboy/simulation/operator/link_info_from_name', self.get_link_info_from_name)
+		self.initial_pose_service = self.create_service(GetLinkPose, '/roboy/simulation/operator/initial_link_pose', self.initial_link_pose_callback)
+
+		p.createConstraint(self.body_id, -1, -1, -1, p.JOINT_FIXED, [0,0,0],[0,0,0],[0,0,0],[0,0,0,1])
```
Comment (missxa) [inline bulletroboy/operator.py:37]: what is this for?

File: bulletroboy/operator.py [hunk 15]
```diff
@@ -3,45 +3,131 @@
...
+
 			link = {}
 			link['name'] = str(p.getJointInfo(self.body_id,i)[12], 'utf-8')
+			link['dims'] = self.get_link_bb_dim(i)
 			link['id'] = i
 			links.append(link)
 		return links
 
 	def get_link_center(self, link_name):
+		"""Gets link's center point in world frame.
```
Comment (missxa) [inline bulletroboy/operator.py:121]: what is a center point? center of mass you mean?

File: bulletroboy/operator.py [hunk 16]
```diff
@@ -3,45 +3,131 @@
...
+		
+		Args:
```
Comment (missxa) [inline bulletroboy/operator.py:123]: incorrect docstring

File: bulletroboy/operator.py [hunk 17]
```diff
@@ -50,17 +136,46 @@ def get_link_index(self, link_name):
 		return index
 	
 	def move(self, case):
+		"""Applies a movement to the operator.
+		
+		Args:
+			case (int): Movement to apply.
```
Comment (missxa) [inline bulletroboy/operator.py:157]: where to find these ints?

File: bulletroboy/operator.py [hunk 18]
```diff
@@ -50,17 +136,46 @@ def get_link_index(self, link_name):
...
 		for ef in ef_names:
+		   #self.get_logger().info('Sending Endeffector pose: ' + ef)
 		   msg = PoseStamped()
 		   ef_id = self.get_link_index(ef)
-		   link_info = p.getLinkState(self.body_id, ef_id)[:2]
+		   link_info = p.getLinkState(self.body_id, ef_id)[4:6]
 		   link_pos = link_info[0]
 		   link_orn = link_info[1]
 
+		   self.drawDebugLines(ef, link_pos)
```
Comment (missxa) [inline bulletroboy/operator.py:198]: why is there drawing in a publisher?

File: bulletroboy/operator.py [hunk 19]
```diff
@@ -74,165 +189,138 @@ def publish_state(self, ef_names = np.array(['human/left_hand','human/right_hand
 
 		   self.ef_publisher.publish(msg)
 
+	def initial_link_pose_callback(self, request, response):
+		self.get_logger().info(f"Service Initial Link Pose: request received for {request.link_name}")
+		head = self.get_link_index(request.link_name)
```
Comment (missxa) [inline bulletroboy/operator.py]: why is this called head?

File: bulletroboy/roboy.py [hunk 20]
```diff
@@ -16,27 +19,46 @@ def __init__(self, body_id):
...
+        # link = p.getLinkState(self.body_id, 7)
+        # jointPoses = p.calculateInverseKinematics(self.body_id, 7, link[0], [0,0,0,1])
+        # for i in range(len(self.freeJoints)):
+        #     p.resetJointState(self.body_id, self.freeJoints[i], jointPoses[i])
+        p.addUserDebugLine([0,0,0],[0.3,0,0],[1,0,0],lineWidth= 3, parentObjectUniqueId=self.body_id, parentLinkIndex=7)
+        p.addUserDebugLine([0,0,0],[0,0.3,0],[0,1,0],lineWidth= 3, parentObjectUniqueId=self.body_id, parentLinkIndex=7)
+        p.addUserDebugLine([0,0,0],[0,0,0.3],[0,0,1],lineWidth= 3, parentObjectUniqueId=self.body_id, parentLinkIndex=7)
+        p.addUserDebugLine([0,0,0],[0.3,0,0],[1,0,0],lineWidth= 3, parentObjectUniqueId=self.body_id, parentLinkIndex=47)
+        p.addUserDebugLine([0,0,0],[0,0.3,0],[0,1,0],lineWidth= 3, parentObjectUniqueId=self.body_id, parentLinkIndex=47)
+        p.addUserDebugLine([0,0,0],[0,0,0.3],[0,0,1],lineWidth= 3, parentObjectUniqueId=self.body_id, parentLinkIndex=47)
```
Comment (missxa) [inline bulletroboy/roboy.py]: can we do this in  a more optimal way?

File: bulletroboy/roboy.py [hunk 21]
```diff
@@ -16,27 +19,46 @@ def __init__(self, body_id):
...
+    
         #Joint state publisher
+        timer_period = 0.1 # seconds
```
Comment (missxa) [inline bulletroboy/roboy.py]: joints must be published more often, at least 50hz, better 100hz

File: bulletroboy/roboy.py [hunk 22]
```diff
@@ -47,39 +69,172 @@ def __init__(self, body_id):
         #Collision publisher
         self.collision_publisher = self.create_publisher(Collision, '/roboy/simulation/collision', 1)
```
Comment (missxa) [inline bulletroboy/roboy.py]:
```suggestion
        self.collision_publisher = self.create_publisher(Collision, '/roboy/simulation/roboy/collision', 1)
```

File: bulletroboy/roboy.py [hunk 23]
```diff
@@ -47,39 +69,172 @@ def __init__(self, body_id):
...
+        """
+        #self.get_logger().info("received call")
+        response.link_name = ""
+        for link in self.get_links():
+            if link['id'] == request.link_id:
+                response.link_name = link['name']
+                response.dimensions.x, response.dimensions.y, response.dimensions.z = link['dims']
+                break
+        #self.get_logger().info("responding")
+        return response
```
Comment (missxa) [inline bulletroboy/roboy.py]: all these fucntions are exact copies of the ones defined for the operator, please optimise and remove copy-paste

File: bulletroboy/roboy.py [hunk 24]
```diff
@@ -47,39 +69,172 @@ def __init__(self, body_id):
...
 
-    def accurateCalculateInverseKinematics(self, targetPos, threshold, maxIter):
-      closeEnough = False
-      iter = 0
-      dist2 = 1e30
-      while (not closeEnough and iter < maxIter):
-        jointPoses = p.calculateInverseKinematics(self.body_id, self.endEffectorId, targetPos)
-        #import pdb; pdb.set_trace()
-        # self.get_logger().info("resetting joints states")
+    def move(self, link_info):
```
Comment (missxa) [inline bulletroboy/roboy.py]: docstrings?

File: bulletroboy/roboy.py [hunk 25]
```diff
@@ -47,39 +69,172 @@ def __init__(self, body_id):
...
+        #self.get_logger().info('Endeffector pose received: ' + link_info.header.frame_id)
 
+        #process message
+        ef_name = OPERATOR_TO_ROBOY_NAMES[link_info.header.frame_id]
```
Comment (missxa) [inline bulletroboy/roboy.py]: this is not using forces/link mapper? if not, it should

Comment (missxa) [reply bulletroboy/roboy.py]: we agreed to delegate all of the mapping to that other node, so neither roboy nor operator simulations should not deal with finding corresponding link names/ids.

File: bulletroboy/roboy.py [hunk 26]
```diff
@@ -47,39 +69,172 @@ def __init__(self, body_id):
...
+
+        #move
+        threshold = 0.001
+        maxIter = 100
+        self.accurateCalculateInverseKinematics(ef_id, link_pos, link_orn, threshold, maxIter)
+        # p.setJointMotorControl2(self.body_id, ef_id, p.POSITION_CONTROL, link_pos)
+        if(ef_name == 'hand_right'):
+            self.drawDebugLine(ef_id, link_pos)
+
+    def accurateCalculateInverseKinematics(self, endEffectorId, targetPos, targetOrn, threshold, maxIter):
```
Comment (missxa) [inline bulletroboy/roboy.py]: add defaults to threshold and maxiter

File: bulletroboy/roboy.py [hunk 27]
```diff
@@ -47,39 +69,172 @@ def __init__(self, body_id):
...
+        #     ls = p.getLinkState(self.body_id, endEffectorId)
+        #     newPos = ls[4]
+        #     diff = [targetPos[0] - newPos[0], targetPos[1] - newPos[1], targetPos[2] - newPos[2]]
+        #     dist2 = (diff[0] * diff[0] + diff[1] * diff[1] + diff[2] * diff[2])
+        #     closeEnough = (dist2 < threshold)
+        #     iter = iter + 1
+        # print("link_info after reset " ,p.getLinkState(self.body_id, endEffectorId)[:2])
+        return jointPoses
+
+    def get_pos_in_world(self, pos, orn):
```
Comment (missxa) [inline bulletroboy/roboy.py]: docstrings missing

also, pos of what?

File: bulletroboy/roboy.py [hunk 28]
```diff
@@ -47,39 +69,172 @@ def __init__(self, body_id):
...
+        torso_pos, torso_orn = p.getBasePositionAndOrientation(self.body_id)
+        return p.multiplyTransforms(torso_pos, torso_orn, pos, orn)
+
+    def adapt_pos_to_roboy(self, pos, orn):
```
Comment (missxa) [inline bulletroboy/roboy.py]: adapt from what to what?

File: bulletroboy/roboy.py [hunk 29]
```diff
@@ -47,39 +69,172 @@ def __init__(self, body_id):
...
+        # rotation = np.array([[0,1,0],[-1,0,0],[0,0,1]])
+        # new_pos = rotation.dot(np.array(pos)) - [0,0.05,0.2]
+
+        # _, new_orn = p.multiplyTransforms(new_pos, new_orn, [0,0,0],self.axis_angle_to_quaternion(0, 0, 1, 90))
+        # new_orn = self.quaternion_multiply(orn, self.axis_angle_to_quaternion(0, 0, 1, -90))
+        # new_orn = self.quaternion_multiply(new_orn, p.getQuaternionFromEuler([0, 0, 1.5708]))
+
+        return new_pos, new_orn
+
+    def axis_angle_to_quaternion(self, ax, ay, az, angle):
```
Comment (missxa) [inline bulletroboy/roboy.py]: another util function candidate , doesnt need to be a method in roboy

**Review by missxa — CHANGES_REQUESTED** [review]

**Comment by missxa** [comment]
a lot of docstrings are missing. also you may remove draft from the PR name

File: bulletroboy/forces_mapper.py (hunk 13 above)
Comment (karim-alweheshy) [reply bulletroboy/forces_mapper.py]: well.. unless someone removes/moves/renames the file it ill work

How do you suggest guaranteeing it?

**Review by missxa — APPROVED** [review]

**Pull request merged**